core/src/main/java/org/apache/calcite/util/format/PostgresqlDateTimeFormatter.java

```diff
@@ -0,0 +1,660 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.util.format;
+
+import java.text.ParsePosition;
+import java.time.Month;
+import java.time.ZonedDateTime;
+import java.time.format.TextStyle;
+import java.time.temporal.ChronoField;
+import java.time.temporal.IsoFields;
+import java.time.temporal.JulianFields;
+import java.util.Locale;
+import java.util.function.Function;
+
+/**
+ * Provides an implementation of toChar that matches PostgreSQL behaviour.
+ */
+public class PostgresqlDateTimeFormatter {
+  /**
+   * Result of applying a format element to the current position in the format
+   * string. If matched, will contain the output from applying the format
+   * element.
+   */
+  private static class PatternConvertResult {
+    final boolean matched;
+    final String formattedString;
+
+    protected PatternConvertResult() {
+      matched = false;
+      formattedString = "";
+    }
+
+    protected PatternConvertResult(boolean matched, String formattedString) {
+      this.matched = matched;
+      this.formattedString = formattedString;
+    }
+  }
+
+  /**
+   * A format element that is able to produce a string from a date.
+   */
+  private interface FormatPattern {
+    /**
+     * Checks if the current position in the format string applies to this format
+     * element. Will produce a formatted string if matched that is the format element
```

Comment:
I find the second phrase difficult to parse,maybe you can rephrase it.

Reply:
I've tried to reword this.